R/domainEnrichment.R

```diff
@@ -0,0 +1,164 @@
+#'
+#' @examples
+#' # Plot enrichend domains in PHALY system orderd by rich factor
+#' # Plot would separate to two parts if too many domains are enriched
+#' # User can hover around the point on plot to get more info (i.e ID descripton)
+#' system <- "PHALY"
+#' domainEnrichment(system)
+#'
+#' @export
+domainEnrichment <- function(system, alpha = 0.05) {
+  # 1. Fetch data from Dr. Steipe's BCB420.2019.ESA package
+  # Two lists from InterPro database
+  genesIPR <- fetchData("genesIPR")
+  IPRgenes <- fetchData("IPRgenes")
+
+  # Genes for the system
+  myDB <- fetchData("SysDB")
+  if (! system %in% names(SyDBgetRootSysIDs(myDB))) {
+    print("Input system should be in database of ESA package.")
+    return(NULL)
+  }
+
+  genes <- SyDBgetSysSymbols(myDB, system)[[1]]
+
+
+  # 2. Find the counts of genes for each domain
+  # 2.1 Store all the domains that in the system and their counts
+  domains <- c()
+  for (gene in genes) {
+    domains <- c(genesIPR[gene][[1]], domains)
+  }
+
+  tmp <- as.data.frame(table(domains))
+  colnames(tmp) <- c("domains", "sysCounts")
+
+  # 2.2 Store the counts from random (all genes)
+  for(i in seq_along(tmp$domains)) {
+    tmp$randomCounts[i] <- length(IPRgenes[as.character(tmp$domains[i])][[1]])
+  }
+
+
+  # 3. Data analysis
+  # 3.1 Calculate the frequency for each domain in the system
+  tmp$sysFreq <- (tmp[,2]) / (sum(genes %in% names(genesIPR)))
+
+  # 3.2 Calculate the frequency for each domain in random (all genes)
+  tmp$randomFreq <- tmp$randomCounts/length(genesIPR)
+
+  # 3.3 Calculate the p_value and odd_ratio
+  # Null hypothesis: The categories are independent.
+  # Some useful page for understanding recommended by Dr. Steipe.
+  # https://www.pathwaycommons.org/guide/primers/statistics/fishers_exact_test/
+  # https://www.pathwaycommons.org/guide/primers/statistics/multiple_testing/
+
+  # Fisher Test Ref:
+  # M. Berk (https://stats.stackexchange.com/users/30201/m-berk), Which
+  # statistical test should be used to test for enrichment of gene lists?,
+  # URL (version: 2013-10-11): https://stats.stackexchange.com/q/72556
+  for (i in seq_along(tmp$domains)) {
+    contingency <- c(tmp$sysCounts[i],
+                     tmp$randomCounts[i],
+                     sum(genes %in% names(genesIPR))-tmp$sysCounts[i],
+                     length(genesIPR)-tmp$randomCounts[i])
+
+    fisherTest <- fisher.test(matrix(contingency, nrow=2, ncol=2),
+                                     alternative="greater")
+
+    tmp$p_value[i] <- fisherTest$p.value
+    tmp$odd_ratio[i] <- fisherTest$estimate
+  }
+
+  # 3.4 Multiple testing: Benjamini-Hochberg control for false discovery rate
+  tmp$p_adj_BH <- p.adjust(tmp$p_value, "BH")
+
+  # 3.5 Calculate the enrichment
+  tmp$enrichment <- (tmp$sysFreq)/(tmp$randomFreq)
+
+
+  # 4. Find the description for enriched domains
+  myMart <- biomaRt::useMart("ensembl", dataset="hsapiens_gene_ensembl")
```

Comment:
That will make the function unbearably slow. I would precompile the domain descriptions. ToDo ...